expression/scalar_function.go

```diff
@@ -165,6 +165,16 @@ func (sf *ScalarFunction) IsCorrelated() bool {
 	return false
 }
 
+// ReferTable implements Expression interface.
+func (sf *ScalarFunction) ReferTable() bool {
+	for _, arg := range sf.GetArgs() {
+		if arg.ReferTable() {
+			return true
+		}
+	}
+	return false
+}
+
```

Comment:
Seems `ExtractColumns` can provide what you want? `CorrelatedColumn` should be treated as constant actually?

Reply:
Oh, my fault. We need to check also non-deterministic functions and statement parameters

Reply:
Oh from [this comment](https://github.com/pingcap/tidb/pull/9751#discussion_r270622898), you need to check if the expression is "purely" constant, i.e, `CorrelatedColumn` should not be treated as constant because we are not in executor, so `ExtractColumns` cannot work, sorry for the misleading.

Why would non-deterministic functions matter here? though their return value is non-deterministic, the return type is deterministic?

What does statement parameters mean?

Reply:
For non-deterministic functions, their return value is non-deterministic, so you can't use its value to computer the frac part or somethings else.
I believe  `statement parameters`  is the parameters used in prepare statement.